cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,441 @@
+        size_t latest_processed_idx = InvalidIdx;
+
+        auto lock = co_await mtx.scoped_lock();
+        co_await request_data.wait(lock, [&] {
+            auto filtered_view = std::ranges::filter_view(ch_next_idx, [](size_t idx) {
+                return idx != InvalidIdx;
+            });
+
+            auto it = std::ranges::begin(filtered_view);  // first valid idx
+            auto end = std::ranges::end(filtered_view);  // end idx
+
+            if (it == end) {
+                // no valid indices, so all send tasks are in an invalid state
+                return true;
+            }
+
+            auto [min_it, max_it] = std::minmax_element(it, end);
+            last_completed_idx = *min_it;
+            latest_processed_idx = *max_it;
+
+            return latest_processed_idx == recv_messages.size();
+        });
+
+        co_return std::make_pair(last_completed_idx, latest_processed_idx);
+    }
+
+    /**
+     * @brief Process input messages and notify send tasks to copy & send messages.
+     *
+     * @param ctx The context to use.
+     * @param ch_in The input channel to receive messages from.
+     * @return A coroutine representing the task.
+     */
+    Node recv_task(Context& ctx, std::shared_ptr<Channel> ch_in) {
+        ShutdownAtExit ch_in_shutdown{ch_in};
+        SetInputDoneAtExit set_input_done{.fanout = this};
+        co_await ctx.executor()->schedule();
+        auto& logger = ctx.logger();
+
+        logger.trace("Scheduled process input task");
+
+        // index of the first message to purge
+        size_t purge_idx = 0;
+
+        // input_done is only set by this task, so reading without lock is safe here
+        while (!input_done) {
+            auto [last_completed_idx, latest_processed_idx] =
+                co_await wait_for_data_request();
+            if (last_completed_idx == InvalidIdx && latest_processed_idx == InvalidIdx) {
+                break;  // all send tasks are in an invalid state, so we need to break
+            }
+
+            // receive a message from the input channel
+            auto msg = co_await ch_in->receive();
+
+            {  // relock mtx to update input_done/recv_messages
+                auto lock = co_await mtx.scoped_lock();
+                if (msg.empty()) {
+                    input_done = true;
+                } else {
+                    recv_messages.emplace_back(std::move(msg));
+                }
+            }
+
+            // notify send_tasks to copy & send messages
+            co_await data_ready.notify_all();
+
+            // purge completed send_tasks. This will reset the messages to empty, so that
+            // they release the memory, however the deque is not resized. This guarantees
+            // that the indices are not invalidated. intentionally not locking the mtx
+            // here, because we only need to know a lower-bound on the last completed idx
+            // (ch_next_idx values are monotonically increasing)
```

Comment:
This comment makes no sense. We're not purging any send tasks, we're just releasing messages that have been handled.

`ch_next_idx` is not touched at all in this loop so the comment about the mutex doesn't seem to make much sense either.